libraries/AP_MSP/AP_MSP_Telem_Backend.cpp

```diff
@@ -348,11 +348,16 @@ void AP_MSP_Telem_Backend::process_incoming_data()
     if (numc > 0) {
         // Process incoming bytes
         while (numc-- > 0) {
-            const uint8_t c = _msp_port.uart->read();
-            msp_parse_received_data(&_msp_port, c);
+			//Read result is signed need to check
+            const int16_t c = _msp_port.uart->read(); 
+			if (c < 0) {
```

Comment:
Did you try without this check? available is checked above and the call is in the same thread as this one

Reply:
Need to format correctly

Reply:
If everything is working correctly the check is not needed. If the uart thread init gets messed up elsewhere the call is needed. Otherwise you get in an infite loop where available says you have data, but you can't actually read it.

Reply:
This is currently broken and not compiling... you made a logic change suggestion and I accepted the change.
The change had a brace matching error...
I've fixed it a pushed to my personal branch... not sure what the process is to get the fix back over here to this pull?

The whole github pull thing is new to me as I have direct write privileges' in the git codebase I work on for a living.
This actually the first pull request I've ever done.....

Reply:
https://ardupilot.org/dev/docs/submitting-patches-back-to-master.html

Reply:
> If everything is working correctly the check is not needed. If the uart thread init gets messed up elsewhere the call is needed. Otherwise you get in an infite loop where available says you have data, but you can't actually read it.

Mmm.... I can't actually see this infinite loop.  `numc` is only assigned to once, and decremented regardless of `read()`'s return value.  Sure, we're probably going to get the same problem next time this method's called, but I don't *think* this is a fall-out-of-sky bug.  Assigning into a `numc` like this is an idiom in our code to avoid the sorts of loop you're thinking of...

I like this top patch - would you mind if I extracted it into a separate PR? (maintaining authorship, of course).

Reply:
Eros saves from crash, but from thread task view it’s an infinite loop running 1024 every 10 msec to no positive effect. I’m away from my computer for next 7 days. So feel free to use any part of this. With your available changes only msp.cop and AP_OSD_MSP_Backend changes needed to fix hdzero ops. Don’t care about anything other than having it functional. Thanks for taking time to review this.